(PCP-732) Refine logging in broker

mruzicka opened the pull request:
This PR updates the log levels of the logged messages according to the following definition:
ERROR: abnormal failure to startup, establish a connection, or deliver a message
WARNING: authorization or authentication failure, or failure to process a message
INFO: change in service or connections (connect/disconnect)
DEBUG: sending or responding to messages with identifying information (but not content), completing state transitions (namely those logged at INFO)
TRACE: receiving messages with full contents, full contents of responses

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -165,15 +165,15 @@
          (send-message connection message))
        (catch Exception e
-         (sl/maplog :debug e
+         (sl/maplog :error e
```
This message happens because the connection is no longer open. Maybe we should make this info, and change the message to something like "Unable to deliver association response because connection is no longer available.".

MikaelSmith replied:
It'd also be nice to incorporate the `requester-uri` in the log message.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -106,10 +106,10 @@
+                      :type :connection-already-associated)
          ;; Logs association status with the following arguments
          ;; 0 : uri associated with incoming connection
          ;; 1 : certificate common name of connection
```
I think we should modify this message a bit to be clear what was done in response: the connection was rejected. So maybe "Rejecting session association for {0} from {1} {2}. Session was already associated as {3}."

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -644,11 +644,11 @@
       #(i18n/trs "Lost connection to controller: {0}" (:uri %)))
     (sl/maplog
-      :debug {:uri uri}
+      :info {:uri uri}
```
Having these two messages one after the other seems odd. I'm not sure the 2nd (removing inventory update subscription) is necessary. We don't reference inventory update subscription anywhere else.

MikaelSmith commented on src/puppetlabs/pcp/broker/shared.clj:
```diff
@@ -116,16 +116,16 @@
         (send-message connection error-msg))
       (catch Exception e
-        (sl/maplog :debug e
+        (sl/maplog :error e
```
This is another case like the error in `process-associate-request!`. I'd be tempted to call this `info`, and change the message to be more clear, something like "Attempted message delivery failed as connection is no longer available."

MikaelSmith replied:
It'd also be nice to incorporate the `requester-uri` in the log message.

puppetcla commented:
CLA signed by all contributors.

mruzicka commented:
@MikaelSmith I've addressed your comments, the travis build failed on the first run, on a re-run it succeeded, the error was not related to the changes in this PR.

MikaelSmith commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -165,15 +165,16 @@
+         (sl/maplog :info e
+                    {:type :message-delivery-error
+                     :target requester-uri}
+                    #(i18n/trs "Error sending associate response to {0}" (:target %)))))
```
It feels incongruous calling this `info` and still saying "Error". Was there something you didn't like about the phrasing I suggested?

MikaelSmith commented on src/puppetlabs/pcp/broker/shared.clj:
```diff
@@ -116,16 +116,17 @@
+        (sl/maplog :info e
+                   {:type :message-delivery-error
+                    :target (:uri connection)}
+                   #(i18n/trs "Error sending error message to {0}" (:target %)))))))
```
As above, I suggested some phrasing that you appear to have ignored. `info` and "Error" are incongruous. I'd like to make that clearer somehow.

MikaelSmith commented:
Similar to the two outstanding suggestions, there are two errors in shared.clj that say "Error in deliver-message" that I think we should clarify (and maybe downgrade to `info`).

mruzicka commented:
@MikaelSmith I've updated the messages according to your suggestion but left out the "connection is no longer available" part as IMO that need not be always the case.
I've also changed the level of the log entries of the four message delivery problems to `:warn`.

adreyer commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -633,7 +634,7 @@
               (maybe-purge-clients! broker timestamp)))
   (sl/maplog
-    :debug {:timeout controller-disconnection-ms}
+    :info {:timeout controller-disconnection-ms}
```
Is this going to make people think things are purged even though they may not be?

mruzicka replied:
I've changed the message to say "potential eviction" and added another log to the function where the eviction actually happens.

mruzicka commented:
I'd go as far as logging the no client certificate case at the warning level as it is an "authorization or authentication failure".

mruzicka commented:
@ipeldridge I've updated the messages concerning the client eviction.

mruzicka commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -508,9 +509,9 @@
              (websockets-client/close! (:websocket old-conn) 4000 (i18n/trs "superseded")))
```
This is not even logged. It's sent to the connection peer as the reason for closing the connection.

mruzicka commented on src/puppetlabs/pcp/broker/core.clj:
```diff
@@ -635,7 +639,7 @@
     :debug {:timeout controller-disconnection-ms}
     ;; 0 : number of milliseconds
-    #(i18n/trs "Scheduled full client eviction in {0,number,integer} ms" (:timeout %))))
+    #(i18n/trs "Scheduled potential full client eviction in {0,number,integer} ms" (:timeout %))))
```
I'd go with the abbreviation as we don't need to care about pluralizing it.
